Bump pycoolmasternet-async and add filter and error code support to CoolMastetNet

amitfin opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
a) New entity attributes to Climate entity of CoolMasterNet:
  1) Clean filter status: boolean indicating if the air filter needs to get cleaned.
  2) Error code: None if no error.

b) Service to allow reseting the clean filter status (once it got cleaned)

Dependency library diff: 
https://github.com/OnFreund/pycoolmasternet-async/compare/v0.1.2...v0.1.5
1) v0.1.2 to v0.1.4 is covered by https://github.com/home-assistant/core/pull/82809
2) v0.1.4 to v0.1.5 has a single change which adds the filter status and error code support.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

amitfin committed:
Add filter and error code support to CoolMastetNet

amitfin commented:
https://user-images.githubusercontent.com/19599059/209467834-f0d8dbf1-4f9b-4dfc-bc7f-3f8e75ab9d1c.mov

amitfin commented:
Comment about the high-level approach:
There is an alternative design of creating separated entities for the additional attributes. However, the CoolMasterNet integration creates non-descriptive unit names be default. The user is requires to manually set the "friendly_name" (or change the "entity_id") for making it more descriptive. If we have an entity per attribute, this manual renaming process can be quite tedious.
In addition, Climate already has multiple attributes, so having 2 more attributes seems reasonable (and has the big advantage of fewer renaming as described above.)
![image](https://user-images.githubusercontent.com/19599059/209474170-9d2be3f6-a3d1-4eae-85b9-a0a56882d0d0.png)

allenporter commented on homeassistant/components/coolmaster/services.yaml:
```diff
@@ -0,0 +1,6 @@
+reset_filter:
+  description: Reset the timer of the filter.
+  fields:
```
service targets are the modern way to do this https://developers.home-assistant.io/docs/documenting/yaml-style-guide?_highlight=target#service-targets e.g. 
```
  target:
    entity:
      integration: coolmaster
      domain: climate
```

allenporter commented on homeassistant/components/coolmaster/climate.py:
```diff
@@ -54,6 +61,9 @@ async def async_setup_entry(
     async_add_devices(all_devices)
 
+    platform = entity_platform.async_get_current_platform()
+    platform.async_register_entity_service(RESET_FILTER, {}, "async_reset_filter")
```
I think the current best practice and modern way to achieve something like this is not to add more to the existing climate entity, but instead add additional entities. For example:
- A sensor or binary sensor entity to show the state of the filter needing to be cleaned
- A button entity to provide a reset
- Additional sensor for the error code state

I just had a look at other integrations doing this and found `sensibo` doing this.

My impression is the motivation behind this is to make it easier for end users to compose these things with other parts of home assistant. For example, the UI for helping with automations. Or by having these entities as `config` or `diagnostic` entity types they show up as options in the UI associated with the underlying device.

allenporter replied:
I see you left a separate comment about this.  I don't think i agree with the rationale as good reasons to go against the current design standards. 

My impression is that the naming issue you describe is handled by giving the device a descriptive names and the current approach is to name entities based on the device name:
https://developers.home-assistant.io/blog/2022/07/10/entity_naming?_highlight=entity&_highlight=name

OnFreund commented:
@allenporter's comments are very similar to the ones I made in https://github.com/OnFreund/pycoolmasternet-async/pull/5. The only difference is that I think the filter state can be represented as a switch rather as a binary sensor+button, but I'm fine either way.

amitfin committed:
Create separate entities

amitfin commented:
Changed the PR according to the recommendation: added separate entities for filter status (binary sensor), rest the filter (button), and error code (sensor).


https://user-images.githubusercontent.com/19599059/209725463-2999e0c9-1906-4bcc-a268-c2c10ed61d78.mov

allenporter commented on homeassistant/components/coolmaster/entity.py:
```diff
@@ -0,0 +1,57 @@
+        super()._handle_coordinator_update()
+
+
+def async_add_entities_for_platform(
```
I'd say just inline this everywhere, duplicating the code for getting the info and coordinator objects and calling the constructor.

amitfin replied:
Done.

amitfin committed:
Remove async_add_entities_for_platform

allenporter commented on homeassistant/components/coolmaster/binary_sensor.py:
```diff
@@ -20,8 +21,13 @@ async def async_setup_entry(
-        hass, config_entry, async_add_entities, CoolmasterCleanFilter
+    info = hass.data[DOMAIN][config_entry.entry_id][DATA_INFO]
+    coordinator = hass.data[DOMAIN][config_entry.entry_id][DATA_COORDINATOR]
+    async_add_entities(
+        [
+            CoolmasterCleanFilter(coordinator, unit_id, info)
+            for unit_id in coordinator.data
+        ]
```
I *think* this works without the conversion to a list here and below.

```suggestion
    async_add_entities(
        CoolmasterCleanFilter(coordinator, unit_id, info)
        for unit_id in coordinator.data
```

amitfin replied:
Done.

allenporter reviewed (commented):
Looks great, one last nit and this is ready to merge

amitfin committed:
Fixed call to async_add_entities

allenporter commented on tests/components/coolmaster/conftest.py:
```diff
@@ -0,0 +1,95 @@
+
+    async def reset_filter(self):
+        """Report that the air filter was cleaned and reset the timer."""
+        TEST_UNITS[self.unit_id]["clean_filter"] = False
```
I was about read to merge and doing a final pass, but then just noticed this is mutating global state, which is going to be a problem. You're going to want to make some kind of data object (or single instance can be shared across the mocks)

amitfin replied:
Fixed. Thanks!

amitfin committed:
Avoid using test global
